### PR TITLE
[#117512517] Check postgres instances deployed via broker are available

### PR DESCRIPTION
**Pull request by HenryTK**
## What

[#117512517 - Check postgres instances deployed via broker are available](https://www.pivotaltracker.com/story/show/117512517)

We want to know whether any changes to our environment affect the service instances and their binding to applications. For this we have deployed a health check application which serves a static endpoint, which replaces the existing health check application, and a database endpoint.

Pingdom health checks can now be set up to monitor the staging and production environments. This should result in the existing health check to be updated, and a new health check created for the database endpoint.

## How to review

* Deploy using `117512517_db_check_app`. 
	* Make sure `cf-deploy` passes.
	* Check The following end points in a browser:
	* `https://healthcheck.${DEPLOY_ENV}.dev.cloudpipelineapps.digital`
	* `https://healthcheck.${DEPLOY_ENV}.dev.cloudpipelineapps.digital/db`
* Set up the Pingdom checks for CI, staging, and production, using the instructions in the README. You can review the contents of the README in the process.
* Log into the Pingdom account and make sure there are checks for both end points for CI, staging and production. Instructions are in the README, which can also be reviewed for clarity.

## Who

Anyone but @henrytk or @combor

### CODE REVIEW

File: concourse/pipelines/create-bosh-cloudfoundry.yml [hunk 1]
```diff
@@ -1235,6 +1209,42 @@ jobs:
...
+                echo | cf login -a ${API_ENDPOINT} -u ${CF_ADMIN} -p ${CF_PASS}
+                cf create-org testers
+                cf set-quota testers test_apps
+                cf create-space healthcheck -o testers
+                cf target -o testers -s healthcheck
+
+                cd paas-cf/tests/example-apps/healthcheck
+                cf push --no-start
+
+                cf create-service postgres M-dedicated-9.5 healthcheck-db
```
Comment (alext) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: This needs to be written in an idempotent way (ie it needs to check if the service already exists before attempting to create it)

File: concourse/pipelines/create-bosh-cloudfoundry.yml [hunk 2]
```diff
@@ -1235,6 +1209,42 @@ jobs:
...
+                while ! cf service healthcheck-db | grep -q 'Status: create succeeded'; do
+                  echo "Waiting for creation of service to complete..."
+                  sleep 30
+                done
+                cf bind-service healthcheck healthcheck-db
```
Comment (alext) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: Similarly, this needs to check whether the binding exists before attempting to create it.

**Comment by alext** [comment]
We need to add something to the destroy pipeline to ensure that the db instance gets destroyed before the CF install is deleted.

File: concourse/pipelines/create-bosh-cloudfoundry.yml (hunk 1 above)
Comment (HenryTK) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: Running the `cf create-service` and `cf bind-service` commands when the services and bindings already exists does not cause a problem. It prints a friendly message saying it does not need to do anything and then exits 0.

**Comment by alext** [comment]
> Remove the existing healthcheck-app from staging and production using the CF CLI.

Can we do this in the pipeline automatically rather than having to have this be a manual step. This is a much more reliable way of doing it.  We can leave it in with a FIXME comment so that we know to remove it later (eg https://github.com/alphagov/paas-cf/commit/8639c6a)

**Comment by HenryTK** [comment]
> Can we do this in the pipeline automatically rather than having to have this be a manual step. This is a much more reliable way of doing it. We can leave it in with a FIXME comment so that we know to remove it later (eg 8639c6a)
We decided it would be less work to just do it as a one-off manual task. Do you think the automated approach would be quicker?

**Comment by dcarley** [comment]
> We decided it would be less work to just do it as a one-off manual task. Do you think the automated approach would be quicker?

It would be preferable. It'll reduce the effort required by the person merging the PR, reduce the chance of it not being applied to an environment (e.g. CI wasn't listed), and increase visibility of how the change made it into our persistent/shared environments.

**Comment by combor** [comment]
> reduce the chance of it not being applied to an environment

At the other hand it increases the chance of not being de-applied, and having leftovers like [this story](https://www.pivotaltracker.com/story/show/118085963) that we actually implemented right now, not realising that binaries were not updated and pingdom checks were left using http: https://github.com/alphagov/paas-cf/pull/295/files#diff-5ae1cbe3159d21d9c29141f9c68cc3bdR19

**Comment by HenryTK** [comment]
@dcarley:
> reduce the chance of it not being applied to an environment (e.g. CI wasn't listed)

I thought we decided not to have checks against CI?

**Comment by dcarley** [comment]
> At the other hand it increases the chance of not being de-applied

Piotr and I talked about this offline. It's alright for the "de-apply" code to be there for a while, since it's effectively a no-op. If we mark it with a `FIXME` comment then it'll be easier to identify and remove. Piotr also suggested requesting a new story in the story summary to remove the FIXME afterwards.

> I thought we decided not to have checks against CI?

Pingdom checks aren't configured to query the environment, but the app is still deployed.

File: concourse/pipelines/autodelete-cloudfoundry.yml [hunk 3]
```diff
@@ -29,6 +50,16 @@ jobs:
...
+      - get: cf-secrets
+      - get: cf-manifest
+      - get: bosh-CA
+
+      - task: get-cf-cli-config
+        file: paas-cf/concourse/tasks/get-cf-cli-config.yml
+
+      - task: remove-healthcheck-app
+        file: paas-cf/concourse/tasks/remove-healthcheck-app.yml
+
```
Comment (keymon) [inline concourse/pipelines/autodelete-cloudfoundry.yml]: Why is the healthcheck-app deleted during the autodelete?

File: concourse/pipelines/destroy-cloudfoundry.yml [hunk 4]
```diff
@@ -103,7 +117,17 @@ jobs:
...
+          - get: bosh-CA
           - get: paas-cf
+          - get: cf-secrets
+          - get: cf-manifest
+
+      - task: get-cf-cli-config
+        file: paas-cf/concourse/tasks/get-cf-cli-config.yml
+
+      - task: remove-healthcheck-app
+        file: paas-cf/concourse/tasks/remove-healthcheck-app.yml
```
Comment (keymon) [inline concourse/pipelines/destroy-cloudfoundry.yml]: As we are deleting CF, the application will get deleted with all the platform, so there is no need to delete it during the destroy.

File: concourse/pipelines/autodelete-cloudfoundry.yml (hunk 3 above)
Comment (combor) [inline concourse/pipelines/autodelete-cloudfoundry.yml]: This not only removes an app itself but also removes RDS instance.

File: concourse/pipelines/create-bosh-cloudfoundry.yml [hunk 5]
```diff
@@ -1258,6 +1232,44 @@ jobs:
 
                   cf push
 
+      - task: deploy-healthcheck
```
Comment (keymon) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: For the sake of consistency and to  make it more clear, shall we move this to a dedicated separated file in tasks? This way we can have a `paas-cf/concourse/tasks/remove-healthcheck-app.yml` and a `paas-cf/concourse/tasks/push-healthcheck-app.yml`

File: concourse/pipelines/autodelete-cloudfoundry.yml (hunk 3 above)
Comment (dcarley) [inline concourse/pipelines/autodelete-cloudfoundry.yml]: Would it be more explicit and faster to only delete the service? Is that possible?

Comment (combor) [inline concourse/pipelines/autodelete-cloudfoundry.yml]: As I remember to delete a service you need to unbind an app. During unbinding the app is restarted so it takes time. Deletion should be quicker. I renamed the task name to `remove-healthcheck-db`.

File: concourse/pipelines/create-bosh-cloudfoundry.yml (hunk 5 above)
Comment (combor) [inline concourse/pipelines/create-bosh-cloudfoundry.yml]: This one is only in `create-bosh-cloudfoundry.yml` so there is no code duplication. I'm not sure about moving it to separate task as we prefer to have tasks inline if possible ( and reasonable ).

**Comment by keymon** [comment]
Everything LGTM, I am about to merge

**Pull request merged**